netebpfext/net_ebpf_ext_hook_provider.c

```diff
@@ -366,33 +474,148 @@ _net_ebpf_extension_hook_provider_attach_client(
+            if (result != EBPF_SUCCESS) {
+                NET_EBPF_EXT_LOG_MESSAGE_UINT32(
+                    NET_EBPF_EXT_TRACELOG_LEVEL_ERROR,
+                    NET_EBPF_EXT_TRACELOG_KEYWORD_EXTENSION,
+                    "net_ebpf_ext_add_client_context failed. Attach attempt rejected.",
+                    result);
+                status = STATUS_ACCESS_DENIED;
+            } else {
+                *provider_binding_context = hook_client;
+                hook_client = NULL;
+            }
+            goto Exit;
+        }
+    } else if (local_provider_context->attach_capability == ATTACH_CAPABILITY_SINGLE_ATTACH_PER_HOOK) {
+        // Exclusive wildcard attach capability. Only one client can be attached with one attach params.
+        // In case of wildcard attach parameter, only one client can be attached overall.
+        if (is_wild_card_attach_parameter) {
+            if (!IsListEmpty(&local_provider_context->filter_context_list)) {
+                NET_EBPF_EXT_LOG_MESSAGE(
+                    NET_EBPF_EXT_TRACELOG_LEVEL_ERROR,
+                    NET_EBPF_EXT_TRACELOG_KEYWORD_EXTENSION,
+                    "Single attach per hook capability. Attach attempt rejected.");
+                status = STATUS_ACCESS_DENIED;
+                goto Exit;
+            }
+        } else {
+            // Check if the attach parameter is already present in the list of filter contexts.
+            net_ebpf_extension_wfp_filter_context_t* matching_context = NULL;
+            matching_context = net_ebpf_extension_get_matching_filter_context(
+                hook_client->client_data->header.size, hook_client->client_data->data, local_provider_context);
+
+            if (matching_context != NULL) {
+                NET_EBPF_EXT_LOG_MESSAGE(
+                    NET_EBPF_EXT_TRACELOG_LEVEL_ERROR,
+                    NET_EBPF_EXT_TRACELOG_KEYWORD_EXTENSION,
+                    "Only one client allowed. Attach attempt rejected.");
+                status = STATUS_ACCESS_DENIED;
+                goto Exit;
+            }
+        }
+    }
+
+    // No matching filter context found. Need to create a new filter context.
+    // Acquire rundown reference on provider context. This will be released when the filter context is deleted.
+    rundown_acquired = ExAcquireRundownProtection(&local_provider_context->rundown.protection);
+    if (!rundown_acquired) {
+        NET_EBPF_EXT_LOG_MESSAGE(
+            NET_EBPF_EXT_TRACELOG_LEVEL_ERROR,
+            NET_EBPF_EXT_TRACELOG_KEYWORD_EXTENSION,
+            "ExAcquireRundownProtection failed. Attach attempt rejected.");
+        status = STATUS_ACCESS_DENIED;
+        goto Exit;
+    }
+
+    // No matching filter context found. Create a new filter context.
+    result = local_provider_context->dispatch.create_filter_context(
+        hook_client, local_provider_context, &new_filter_context);
+    if (result != EBPF_SUCCESS) {
```

Comment:
why not call `net_ebpf_ext_add_client_context` instead from inside `net_ebpf_extension_wfp_filter_context_create`?

Reply:
`net_ebpf_ext_add_client_context()` is only used for the case where the filter context already exists and a new client is being appended to the list of clients.